### PR TITLE
Add an AMF Deserializer as well as corresponding Unit Tests

### PR DESCRIPTION
**Pull request by EmosewaMC**
This PR adds an AMF Deserializer written by myself specifically for DLU to be used for property behavior GameMessages.  The property behaviors themselves are not touched in the slightest with this PR however this is the foundation for properly processing the information the client sends to the server for the needed actions to be done.  Working on implementing the rest of the unit test and also implementing some bounds checks.

Unit tests are attached and test that the deserialized AMF matches one sent from the live game servers.

### CODE REVIEW

**Comment by EmosewaMC** [comment]
@Xiphoseer Could you give some feedback on the tests I have running against the deserializer and let me know if i'm missing anything or if it looks like ive got the edge cases covered pretty well?

**Review by Xiphoseer — COMMENTED** [review]
Looks good overall. Would have expected the read function(s) to follow the grammar in the spec a bit more, though. And there seems to be a distinct lack of U29 decoding, which leads me to assume that it'd read nonsense for arrays with length > 128?

Can you implement that / add a test for it?

File: dCommon/AMFDeserialize.cpp [hunk 1]
```diff
@@ -0,0 +1,151 @@
...
+ */
+AMFArrayValue* AMFDeserialize::Read(RakNet::BitStream* inStream, bool readStartOfArray, std::string parentKey) {
+    if (!inStream) return nullptr;
+    AMFArrayValue* values = new AMFArrayValue();
+    if (readStartOfArray) {
+        int8_t start;
+        int8_t size;
+        inStream->Read(start);
+        inStream->Read(size);
+    }
```
[the remark this answers is not in the document]
Comment (EmosewaMC) [reply dCommon/AMFDeserialize.cpp:18]: The intent of this section is to read what would be in binary 09 01 from the start of an array.  This probably should be adjusted to follow the file spec a bit closer since this portion of any amf was generally always started with 09 01.

File: dCommon/AMFDeserialize.cpp [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+                int8_t character;
+                inStream->Read(character);
+                key.push_back(character);
+            }
+            // Empty strings are never sent by reference.  This should never be empty here though...
+            if (!key.empty()) accessedElements.push_back(key);
+        } else {
+            key = accessedElements[sizeOfString];
+        }
+        // Read in the value type from the bitStream
```
[the remark this answers is not in the document]
Comment (EmosewaMC) [reply dCommon/AMFDeserialize.cpp:44]: That sounds like a good way to keep work more separate.

File: dCommon/AMFDeserialize.cpp [hunk 3]
```diff
@@ -0,0 +1,151 @@
+#include "AMFDeserialize.h"
+
+#include "AMFFormat.h"
+
+/**
+ * AMF3 Reference document https://rtmp.veriskope.com/pdf/amf3-file-format-spec.pdf
+ * I have only written the values that the LEGO Universe client sends to the server.  All others are left out.
+ * AMF3 Deserializer written by EmosewaMC
+ */
+AMFArrayValue* AMFDeserialize::Read(RakNet::BitStream* inStream, bool readStartOfArray, std::string parentKey) {
```
[the remark this answers is not in the document]
Comment (EmosewaMC) [reply dCommon/AMFDeserialize.cpp:10]: interesting.  Ill look at making the method static and also enhancing the deletion process of amf arrays.  currently they dont have any destructors at all which could cause leaks since those values may be forgotten about.  Ill look at the unique_ptr as well.  Not something Im too familiar with but ill give it a go regardless

File: tests/AMFDeserializeTests.cpp [hunk 4]
```diff
@@ -0,0 +1,286 @@
...
+			ReadFromBitStream(&testBitStream);
+		} catch (AMFValueType unimplementedValueType) {
+			caughtException = true;
+		}
+		std::cout << "Testing unimplemented value " << amfValueType << " Did we catch an exception: " << (caughtException ? "YES" : "NO") << std::endl;
+		assert(caughtException);
+	}
+}
+
+void TestLoad() {
```
[the remark this answers is not in the document]
Comment (EmosewaMC) [reply tests/AMFDeserializeTests.cpp:64]: The goal of this test was to make sure the time it would take to parse an extraordinarily long packet would not exceed the maximum time we would want the server to spent parsing packets.  If the test is not useful it can be removed.

File: tests/AMFDeserializeTests.cpp [hunk 5]
```diff
@@ -0,0 +1,286 @@
...
+	// Read a test BitStream from a file
+	RakNet::BitStream bitStream;
+	char byte = 0;
+	while (fileStream.get(byte)) {
+		bitStream.Write<char>(byte);
+	}
+
+	fileStream.close();
+
+	ReadFromBitStream(&bitStream);
```
[the remark this answers is not in the document]
Comment (EmosewaMC) [reply tests/AMFDeserializeTests.cpp:79]: All of these tests likely need to implement the proper memory management.  Since the AMF format we use does not properly delete stuff I did not implement any memory management since you would need to access every possible node here to delete them and thats not what the test does.  As mentioned in another reply, the delete for AMF format will be enhanced to properly manage memory when delete is called on it so it deletes all possible and existing associative maps and dense values.

File: tests/AMFDeserializeTests.cpp [hunk 6]
```diff
@@ -0,0 +1,286 @@
...
+	// unimplemented AMF values are attempted to be parsed.
+	std::ifstream fileStream;
+	fileStream.open("AMFBitStreamUnimplementedTest.bin", std::ios::binary);
+
+	// Read a test BitStream from a file
+	std::vector<char> baseBitStream;
+	char byte = 0;
+	while (fileStream.get(byte)) {
+		baseBitStream.push_back(byte);
+	}
```
[the remark this answers is not in the document]
Comment (EmosewaMC) [reply tests/AMFDeserializeTests.cpp:43]: Ill investigate some better ways to read files. Seemed good for getting the base tests across for now though.

**Comment by EmosewaMC** [comment]
> Looks good overall. Would have expected the read function(s) to follow the grammar in the spec a bit more, though. And there seems to be a distinct lack of U29 decoding, which leads me to assume that it'd read nonsense for arrays with length > 128?
> 
> Can you implement that / add a test for it?

This would be the correct assumption.  I had completely forgotten that _any_ number can be a U29 (or signed 29) depending on where they are.  Ill create tests for this after implementing them and applying the feedback given above

File: dCommon/AMFDeserialize.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (Wincent01) [reply dCommon/AMFDeserialize.cpp:10]: AMF and LDF should both ideally utilize smart pointers, as deletion is seldom consistent and **does** cause sizable memory leaks. Smart pointers aren't used anywhere else in the codebase, so would require quite a bit of refactoring to introduce them. Would suggest a separate PR for that though.

[the remark this answers is not in the document]
Comment (EmosewaMC) [reply dCommon/AMFDeserialize.cpp:10]: Interesting.  Since this code isnt going to be used in the live game until I implement the case in the message handler, ill possibly put that in before I put in the actual property behavior stuff to continue cleaning up the memory since any user could create lots of memory leaks very quickly with large AMFs the client can send.

**Comment by Xiphoseer** [comment]
For historical reference: When AMF was write-only, you could in most cases allocate all values on the stack and then pass references to the array/object constructors and have it destructed on its own. If we move to smart pointers, ot might make sense to use a ptr + `is_owned` combination smart pointer that only calls delete if the latter is true.

**Comment by Xiphoseer** [comment]
(because unique_ptr always allocates and shared_ptr has the ref counting overhead + allocation)

File: dCommon/AMFDeserialize.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (Xiphoseer) [reply dCommon/AMFDeserialize.cpp:18]: FWIW, 0x09 is the array marker, which should not neer to be special cased. It's just that the root value can be any type.

[the remark this answers is not in the document]
Comment (EmosewaMC) [reply dCommon/AMFDeserialize.cpp:18]: I hadn't thought about the fact that amf could be sent not contained within a base array.

[the remark this answers is not in the document]
Comment (Xiphoseer) [reply dCommon/AMFDeserialize.cpp:18]: Yeah, I think it helps to compare it to JSON: `0`, `true`, `"hi"` are all valid json documents even though most APIs will produce something wrapped in `[...]` or `{...}`

File: dCommon/AMFDeserialize.h [hunk 7]
```diff
@@ -0,0 +1,21 @@
...
+#include <string>
+
+class AMFArrayValue;
+class AMFDeserialize {
+    public:
+        /**
+         * Read an AMF3 value from a bitstream.
+         * 
+         * @param inStream inStream to read value from.
+         * @param doFullProcess currently this is here for arrays but should be removed and have its work be done in the AMFArray switch case.
```
Comment (Marcono1234) [inline dCommon/AMFDeserialize.h:15]: The parameter `doFullProcess` does not exist; should that be `readStartOfArray`?

File: dCommon/AMFDeserialize.cpp [hunk 8]
```diff
@@ -0,0 +1,151 @@
...
+                throw static_cast<AMFValueType>(valueType);
+                break;
+        }
+        // Read size of next string
+        inStream->Read(sizeOfString);
+
+        if (sizeOfString % 2 == 0) isReference = true;
+        else isReference = false;
+
+        sizeOfString = sizeOfString >> 1;
```
Comment (Marcono1234) [inline dCommon/AMFDeserialize.cpp:148]: Would it make sense to change the loop to `while (true) { ... break; ... }` to only have this logic for reading and decoding the size once at the start of the loop, instead of it being duplicated before and at the end of the loop here?

Would it maybe also make sense to rename the variable `sizeOfString` to `sizeOfKey` / `sizeOfKeyString` to avoid confusion with a string value?

File: dCommon/AMFDeserialize.h [hunk 9]
```diff
@@ -0,0 +1,21 @@
...
+         * @return Returns an AMFArrayValue with all the information from the bitStream in it.
+         */
+        AMFArrayValue* Read(RakNet::BitStream* inStream, bool readStartOfArray = false, std::string parentKey = "");
```
Comment (Marcono1234) [inline dCommon/AMFDeserialize.h:18]: What is `parentKey` used for?

File: dCommon/AMFDeserialize.cpp [hunk 10]
```diff
@@ -0,0 +1,151 @@
...
+                    for (uint32_t i = 0; i < length; i++) {
+                        int8_t character;
+                        inStream->Read(character);
+                        value.push_back(character);
+                    }
+                    // Empty strings are never sent by reference
+                    if (!value.empty()) accessedElements.push_back(value);
+                } else {
+                    // Length is a reference!
+                    length = length / 2;
```
Comment (Marcono1234) [inline dCommon/AMFDeserialize.cpp:98]: Should this for consistency with the other lines here also be `length >> 1` instead?

File: tests/AMFDeserializeTests.cpp (hunk 6 above)
[the remark this answers is not in the document]
Comment (Jettford) [reply tests/AMFDeserializeTests.cpp:43]: @Xiphoseer Isn't this probably the best way for getting it into a vector of uint8s though?

[the remark this answers is not in the document]
Comment (Xiphoseer) [reply tests/AMFDeserializeTests.cpp:43]: no because it's just semantically the same (which, granted, a compiler might recognize and optimize out), but what you really want is

1. Choose a power of 2 for chunks to read at once
2. Allocate a vector with the length of the input (which you can get for a file by seeking to end and getting the offset), possibly rounded up to a multiple of your chunk size
3. Ask the file stream to read one chunk of bytes at once in a loop, which will just produce less for the last iteration.

The underlying OS is usually really good at putting a bunch of sequential bytes in a buffer with one system call, no need trigger those with increments of one by reading one byte at a time from the file stream.

[the remark this answers is not in the document]
Comment (Xiphoseer) [reply tests/AMFDeserializeTests.cpp:43]: Additionally, if you go from buffer to buffer, C functions like `memcpy` already take care of optimizing small, large and (un)aligned copies efficiently.

File: tests/AMFDeserializeTests.cpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (EmosewaMC) [reply tests/AMFDeserializeTests.cpp]: Addressed

File: dCommon/AMFDeserialize.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (EmosewaMC) [reply dCommon/AMFDeserialize.cpp]: Addressed this for the tests - unique_ptr re-write for AMF will need to wait for another PR.

File: dCommon/AMFDeserialize.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (EmosewaMC) [reply dCommon/AMFDeserialize.cpp]: Addressed

**Comment by EmosewaMC** [comment]
This is ready for re-review.  Those who left feedback feel free to do so again

File: dCommon/AMFDeserialize.cpp [hunk 11]
```diff
@@ -0,0 +1,163 @@
...
+	} else {
+		while (true) {
+			auto key = ReadString(inStream);
+			// No more values when we encounter an empty string
+			if (key.size() == 0) break;
+			arrayValue->InsertValue(key, Read(inStream, false));
+		}
+	}
+	return arrayValue;
+}
```
Comment (EmosewaMC) [inline dCommon/AMFDeserialize.cpp:151]: Why exactly is this an if else read from the bitstream and not a read dense then read associative map?  Perhaps I am missing something on why its needed to be written like this.

**Review by Xiphoseer — COMMENTED** [review]
LGTM overall, minus the way `accessedElements` is a static (ie global) variable. That makes the whole deserializer not threadsafe for no good reason.

That's an "easy" fix though: remove `static` everywhere, remove `clearKeys`. Then just add a `AMFDeserialize deserializer;` in tests wherever you need one and replace `AMFDeserialize::` with `deserializer->`.

Most of the time that'll just be an empty std::vector sitting on the stack, and one additional pointer sitting in a register somewhere.

The keys will be cleared automatically when the `deserializer` instance is deconstructed and if you really want to be able to do multiple parses with the same instance, you can add a manual `clear` method.

File: .gitignore [hunk 12]
```diff
@@ -119,4 +119,6 @@ thirdparty/zlib-1.2.11/
 .env
 docker/__pycache__
 
-docker-compose.override.yml
\ No newline at end of file
+docker-compose.override.yml
+
+!/tests/TestBitStreams/*.bin
```
[the remark this answers is not in the document]
Comment (EmosewaMC) [reply .gitignore:124]: Yes

File: dCommon/AMFDeserialize.h [hunk 13]
```diff
@@ -0,0 +1,71 @@
...
+		 * 
+		 * @param inStream bitStream to read data from
+		 * @return Integer value represented as an AMFValue
+		 */
+		static AMFValue* ReadAmfInteger(RakNet::BitStream* inStream);
+
+		/**
+		 * List of strings read so far saved to be read by reference.
+		 */
+		static std::vector<std::string> accessedElements;
```
[the remark this answers is not in the document]
Comment (EmosewaMC) [reply dCommon/AMFDeserialize.h:70]: I made this static because of your comment earlier in the thread lol.  I'll make this not static anymore.

File: dCommon/AMFDeserialize.cpp [hunk 14]
```diff
@@ -0,0 +1,163 @@
...
+			returnValue = ReadAmfString(inStream);
+			break;
+		}
+
+		case AMFValueType::AMFArray: {
+			returnValue = ReadAmfArray(inStream);
+			break;
+		}
+
+		// Don't think I need these for now.  Will throw if one is encountered since it would break a de-serialization.
```
[the remark this answers is not in the document]
Comment (EmosewaMC) [reply dCommon/AMFDeserialize.cpp:60]: Sure

File: dCommon/AMFDeserialize.cpp [hunk 15]
```diff
@@ -0,0 +1,163 @@
...
+			inStream->Read(character);
+			returnString += character;
+		}
+		// Empty strings are never sent by reference
+		if (!returnString.empty()) accessedElements.push_back(returnString);
+	} else {
+		// Length is a reference to a previous index - use that as the read in value
+		returnString = accessedElements[length];
+	}
+	return returnString;
```
[the remark this answers is not in the document]
Comment (EmosewaMC) [reply dCommon/AMFDeserialize.cpp:122]: It does not.  The amf spec specifically says that empty strings are never sent by reference.

File: .gitignore (hunk 12 above)
[the remark this answers is not in the document]
Comment (Xiphoseer) [reply .gitignore:124]: Perfect!

File: dCommon/AMFDeserialize.h (hunk 13 above)
[the remark this answers is not in the document]
Comment (Xiphoseer) [reply dCommon/AMFDeserialize.h:70]: Right, you mean https://github.com/DarkflameUniverse/DarkflameServer/pull/599#discussion_r906762914m, I presume. Sorry about that, your original design there was fine in that respect. I guess my emphasis was meant to be more on *method on `AMFArrayValue`* (which has to be static because you don't have an instance yet), rather than on *consider making this [..] static*. Should have caught that earlier.

File: dCommon/AMFDeserialize.cpp (hunk 15 above)
[the remark this answers is not in the document]
Comment (EmosewaMC) [reply dCommon/AMFDeserialize.cpp]: Opting to use &value[0] since stl specifies to use this for a modifiable char*.
